reference_data/management/tests/update_gencode_tests.py

```diff
@@ -0,0 +1,207 @@
+
+        # Test the code paths of generating urls, gencode_release <= 22
+        mock_logger.reset_mock()
+        call_command('update_gencode', '--gencode-release=20')
+        mock_download.assert_called_with("http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_20/gencode.v20.annotation.gtf.gz")
+
+        # Test the code paths of generating urls, gencode_release > 22
+        mock_logger.reset_mock()
+        mock_download.return_value = self.temp_file_path
+        call_command('update_gencode', '--gencode-release=23')
+        calls = [
+            mock.call("http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_23/GRCh37_mapping/gencode.v23lift37.annotation.gtf.gz"),
+            mock.call(
+                "http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_23/gencode.v23.annotation.gtf.gz"),
+        ]
+        mock_download.assert_has_calls(calls)
+
+    @mock.patch('reference_data.management.commands.update_gencode.logger')
+    def test_update_gencode_command(self, mock_logger):
+        # Test normal command function
+        call_command('update_gencode', '--gencode-release=31', self.temp_file_path, '37')
+        calls = [
+            mock.call(
+                'Loading {} (genome version: 37)'.format(self.temp_file_path)),
+            mock.call('Creating 1 GeneInfo records'),
+            mock.call('Creating 2 TranscriptInfo records'),
+            mock.call('Done'),
+            mock.call('Stats: '),
+            mock.call('  genes_skipped: 1'),
+            mock.call('  genes_created: 1'),
+            mock.call('  transcripts_created: 2')
+        ]
+        mock_logger.info.assert_has_calls(calls)
+
+        gene_infos = {gene.gene_id: {
+            'start_grch37': gene.start_grch37,
+            'chrom_grch37': gene.chrom_grch37,
+            'coding_region_size_grch37': gene.coding_region_size_grch37,
+            'gencode_release': gene.gencode_release,
+            'gencode_gene_type': gene.gencode_gene_type,
+            'gene_id': gene.gene_id,
+            'gene_symbol': gene.gene_symbol,
+            'end_grch37': gene.end_grch37,
+            'strand_grch37': gene.strand_grch37
+        } for gene in GeneInfo.objects.filter(gene_id__in = ['ENSG00000223972', 'ENSG00000284662'])}
+        self.assertDictEqual(gene_infos['ENSG00000223972'], {'start_grch37': 11869, 'chrom_grch37': u'1', 'coding_region_size_grch37': 0, 'gencode_release': 27, 'gencode_gene_type': u'transcribed_unprocessed_pseudogene', 'gene_id': u'ENSG00000223972', 'gene_symbol': u'DDX11L1', 'end_grch37': 14409, 'strand_grch37': u'+'})
+        self.assertDictEqual(gene_infos['ENSG00000284662'], {'start_grch37': 621059, 'chrom_grch37': u'1', 'coding_region_size_grch37': 936, 'gencode_release': 31, 'gencode_gene_type': u'protein_coding', 'gene_id': u'ENSG00000284662', 'gene_symbol': u'OR4F16', 'end_grch37': 622053, 'strand_grch37': u'-'})
```

Comment:
This is very hard to read, please do this line by line. Also, the script only created 1 gene - why are you testing 2?

Reply:
To ensure the skipped one hasn't been changed.